app/services/blacklight/search_service.rb

```diff
@@ -0,0 +1,120 @@
+      response = repository.search(builder)
+
+      if response.grouped? && grouped_key_for_results
+        [response.group(grouped_key_for_results), []]
+      elsif response.grouped? && response.grouped.length == 1
+        [response.grouped.first, []]
+      else
+        [response, response.documents]
+      end
+    end
+
+    # retrieve a document, given the doc id
+    # @param [Array{#to_s},#to_s] id
+    # @return [Blacklight::Solr::Response, Blacklight::SolrDocument] the solr response object and the first document
+    def fetch(id = nil, paging_params = {}, extra_controller_params = {})
+      if id.is_a? Array
+        fetch_many(id, paging_params, extra_controller_params)
+      else
+        fetch_one(id, extra_controller_params)
+      end
+    end
+
+    ##
+    # Get the solr response when retrieving only a single facet field
+    # @return [Blacklight::Solr::Response] the solr response
+    def facet_field_response(facet_field, user_params, extra_controller_params = {})
+      query = search_builder.with(user_params).facet(facet_field)
+      repository.search(query.merge(extra_controller_params))
+    end
+
+    # Get the previous and next document from a search result
+    # @return [Blacklight::Solr::Response, Array<Blacklight::SolrDocument>] the solr response and a list of the first and last document
+    def previous_and_next_documents_for_search(index, request_params, extra_controller_params = {})
+      p = previous_and_next_document_params(index)
+      query = search_builder.with(request_params).start(p.delete(:start)).rows(p.delete(:rows)).merge(extra_controller_params).merge(p)
+      response = repository.search(query)
+      document_list = response.documents
+
+      # only get the previous doc if there is one
+      prev_doc = document_list.first if index > 0
+      next_doc = document_list.last if (index + 1) < response.total
+      [response, [prev_doc, next_doc]]
+    end
+
+    # a solr query method
+    # does a standard search but returns a simplified object.
+    # an array is returned, the first item is the query string,
+    # the second item is an other array. This second array contains
+    # all of the field values for each of the documents...
+    # where the field is the "field" argument passed in.
+    def opensearch_response(request_params, field = nil, extra_controller_params = {})
+      field ||= blacklight_config.view_config(:opensearch).title_field
+
+      query = search_builder.with(request_params).merge(solr_opensearch_params(field)).merge(extra_controller_params)
+      response = repository.search(query)
+
+      [response.params[:q], response.documents.flat_map { |doc| doc[field] }.uniq]
+    end
+
+    ##
+    # The key to use to retrieve the grouped field to display
+    def grouped_key_for_results
+      blacklight_config.index.group
+    end
+
+    delegate :repository_class, to: :blacklight_config
+
+    def repository
+      repository_class.new(blacklight_config)
```

Comment:
I think we also have some use cases for varying the `repository` based on user parameters (previously supported by overriding the method in the controller). I don't think this architecture necessarily precludes that (especially if we can pull the user params up to an attribute), but it would be good to think that through more.